JavaScript: Add has* methods

seishun opened the pull request:
As discussed in #1447, this adds methods that return presence information for non-repeated proto2 fields.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok to test.

haberman commented:
Thanks a lot for writing this up. I think this is the right approach, and we should move forward with it.

Unfortunately, for substantial, API-impacting changes like this, we need to make the change internally first. This is how we ensure that internal code can be migrated and that there aren't unforeseen issues that come up. I'm working on doing this internally right now. I'll let you know as this progresses.

seishun commented:
Rebased on master.

seishun commented:
The test failures seem unrelated.

haberman commented:
I wanted to let you know that we just merged a change from Google-internal development that does the same thing as your PR. So we can close this PR now.

However we can still use your fixes in https://github.com/google/protobuf/pull/1447, since we have not implemented these yet.

seishun commented:
@haberman I noticed that 3.0.0-beta-4 doesn't implement has* methods for message fields. Is that intended?

haberman commented:
Sort of. I implemented it that way to match the current behavior of clear*, which doesn't exist for message fields. That said, I don't think there should be any issue adding clear* and has* for message fields.

haberman commented:
Are you updating your other PR right now? I ask because we're in a bit of a time crunch and trying to get the presence fixes in sometime next week. If that timeframe doesn't work for you, I'll probably adopt your PR and try to get it merged myself.

seishun commented:
>I implemented it that way to match the current behavior of clear*, which doesn't exist for message fields.

Message fields do have clear* methods though. See [here](https://github.com/google/protobuf/blob/9086d9643903c608ab015b0b7d903547a4e7b6f3/src/google/protobuf/compiler/js/js_generator.cc#L2341-L2350).

>Are you updating your other PR right now?

Yes, I've already rebased it onto 3.0.0-beta-4 and replaced the old presence checks (`jspb.Message.getField(msg, $index$) != null`) with has* method calls. That's when I realized they don't exist for message fields.

>I ask because we're in a bit of a time crunch and trying to get the presence fixes in sometime next week. If that timeframe doesn't work for you, I'll probably adopt your PR and try to get it merged myself.

I could easily get this done today. I'm just not sure if I should replace all has* calls with the old presence checks, or just for message fields. I hope to avoid this dilemma by having has* methods for message fields too.

haberman commented:
I think adding `has*` methods for message fields should be fine. Thanks and I'll look forward to seeing your rebased PR!

seishun commented:
Should I submit a PR for `has*` methods for message fields, or will you do it?

haberman commented:
How about making it a part of your PR for presence?
